Some bug fixing, cleanup and shared file

Thaodan opened the pull request:
This pr fixes some lighther bugs like quoting or legacy backtics and moves everything that can be shared between the two scripts into a shared file.

MarijnS95 committed:
clang: Update for Q.

Thaodan committed:
Move everything that can be shared into shared files

The first contains the logic and the second all variables to be
sourced before.

Signed-off-by: Björn Bidar <theodorstormgrade@gmail.com>

Thaodan committed:
Use $PWD rather than calling an external command

Signed-off-by: Björn Bidar <theodorstormgrade@gmail.com>

Thaodan committed:
Fix quoting to prevent word splitting.

Signed-off-by: Björn Bidar <theodorstormgrade@gmail.com>

Thaodan committed:
Just exit on error instead of trying to parse something.

Signed-off-by: Björn Bidar <theodorstormgrade@gmail.com>

Thaodan committed:
Use $() instead of legacy backticket `...`

See also:
https://github.com/koalaman/shellcheck/wiki/Sc2006

Signed-off-by: Björn Bidar <theodorstormgrade@gmail.com>

oshmoun commented:
while you're at it, it would be invaluable to have the actual build script seperate from the iteration over the devices to build. that way one has a simple script to build any device of their choice.

MarijnS95 commented:
@Thaodan Thanks for PR-ing this! I was about to ask in #4 whether to fix-up my commits and PR those, looks like you already took care of it.

I'll submit my cleaned-up (not yet pushed) patch tomorrow that takes care of the following (and rebased on this shared approach which is a really good idea :+1:):
- Build single devices
- To their own folder
- Does not delete output folder unless you ask for it (incremental builds, which are necessary for everyone without a ridiculous 56-core machine)

Pretty much what you just asked @oshmoun !

Thaodan commented:
I wanted to the latter later because I already had so much cleanup done and add another PR for the rest. 
I don't think in their own folder build is needed. When building multiple times ccache should be used and when I build for one device I don't need to remove the folder before starting a new build.

MarijnS95 commented:
@Thaodan Yes, that is generally true. The build system will figure out what to rebuild and what to link into the final vmlinux image, so that is safe to do.
Separate directories can be made optional, but I like it since disk space is more readily available on my build system in contrast to CPU speed (== lost time). Even with ccache it still has to go over each file, which takes almost a minute on my machine where a rebuild only takes 10-20 seconds, or something like that.

> and add another PR for the rest.

Does that mean you will be submitting changes to do a per-device build easily?
(Meaning both _within_ an aosp environment as well as outside it, by eg. specifying the device as argument or env var)

Anyway, I heard @ix5 is looking into submitting the Android-10 compatible build scripts which could make this all a non-issue, as long as it doesn't involve too much hacking around the build system itself to get it to a somewhat user-friendly state.

ix5 commented on build-kernels-clang.sh:
```diff
@@ -9,6 +9,7 @@ export CLANG_CC=$ANDROID_ROOT/prebuilts/clang/host/linux-x86/clang-r353983c/bin/
 
 # Mkdtimg tool
 export MKDTIMG=$ANDROID_ROOT/out/host/linux-x86/bin/mkdtimg
+export MKDTIMG=$ANDROID_ROOT/prebuilts/misc/linux-x86/libufdt/mkdtimg
```
Duplicate, delete the previous line

MarijnS95 replied:
Yeah I forgot to add a comment, `MKDTIMG` and `CP_BLOB` can be dropped since these are now defined in the generic script.

ix5 commented on build-kernels-clang.sh:
```diff
@@ -17,96 +17,5 @@ export BUILD="make O=$KERNEL_TMP ARCH=arm64 CC=$CLANG_CC CLANG_TRIPLE=aarch64-li
-unset KERNEL_TMP
-unset BUILD
+# source shared parts
+. "$OLDPWD"/build_shared.sh
```
`OLDPWD` is a bit too much magic, why not use `dirname $0` so this script can be called from anywhere?

Thaodan replied:
$OLDPWD is a builtin var why use dirname here? There's no reason to call an external command.

ix5 reviewed (changes requested):
Thanks for taking a dig at this!

oshmoun commented:
one more suggestion (yes i know I'm just pestering at this point sorry!) but can logging be done with `tee` so that it still outputs to console in addition to a file?

Thaodan commented:
It can and I thought of that too but I 
thought the point of that of output this 
wasn't intended.

MartinX3 commented:
Sadly I can't build tama on my Ubuntu 18.04 with your PR
(It works with this https://github.com/MarijnS95/kernel-sony-msm-4.14-common/commits/aosp/LA.UM.7.1.r1 commits so my buildmachine is able to build the kernel and DTBO's)

Or maybe you need to rebase the PR, because I simply downloaded your branch and replaced the files on my computer with your files.

I also did remove the kernel-tmp folder before trying this PR.

```
=================================================
Your Environment:
ANDROID_ROOT: /home/developer/android/rom/sonyAOSP/10
KERNEL_TOP  : /home/developer/android/rom/sonyAOSP/10/kernel/sony/msm-4.14
KERNEL_TMP  : /home/developer/android/rom/sonyAOSP/10/out/kernel-tmp
=================================================
Platform -> tama :: Device -> akari
make[1]: Verzeichnis „/media/martin/extLinux/developer/android/cache/sonyAOSP/10/kernel-tmp“ wird betreten
  GEN     ./Makefile
  HOSTCC  scripts/basic/fixdep
  HOSTCC  scripts/kconfig/conf.o
  SHIPPED scripts/kconfig/zconf.tab.c
  SHIPPED scripts/kconfig/zconf.lex.c
  HOSTCC  scripts/kconfig/zconf.tab.o
  HOSTLD  scripts/kconfig/conf
#
# configuration written to .config
#
make[1]: Verzeichnis „/media/martin/extLinux/developer/android/cache/sonyAOSP/10/kernel-tmp“ wird verlassen
The build may take up to 10 minutes. Please be patient ...
Building new kernel image ...
Logging to /home/developer/android/rom/sonyAOSP/10/out/kernel-tmp/build_log_akari
Copying new kernel image ...
create image file: /home/developer/android/rom/sonyAOSP/10/kernel/sony/msm-4.14/common-kernel/dtbo-akari.img...
Can not read file: /home/developer/android/rom/sonyAOSP/10/out/kernel-tmp/arch/arm64/boot/dts/qcom/sdm845-v2.1-tama-akari_generic-overlay.dtbo
/home/developer/android/rom/sonyAOSP/10/out/kernel-tmp/arch/arm64/boot/dts/qcom/sdm845-tama-akari_generic-overlay.dtbo
/home/developer/android/rom/sonyAOSP/10/out/kernel-tmp/arch/arm64/boot/dts/qcom/sdm845-v2-tama-akari_generic-overlay.dtbo
Can not output image with args
```

The mentioned files do exist
```
ls -lha /home/developer/android/rom/sonyAOSP/10/out/kernel-tmp/arch/arm64/boot/dts/qcom/sdm845-v2.1-tama-akari_generic-overlay.dtbo
-rw-r--r-- 1 martin martin 291K Nov 22 18:05 /home/developer/android/rom/sonyAOSP/10/out/kernel-tmp/arch/arm64/boot/dts/qcom/sdm845-v2.1-tama-akari_generic-overlay.dtbo
```
```
ls -lha /home/developer/android/rom/sonyAOSP/10/out/kernel-tmp/arch/arm64/boot/dts/qcom/sdm845-tama-akari_generic-overlay.dtbo
-rw-r--r-- 1 martin martin 291K Nov 22 18:05 /home/developer/android/rom/sonyAOSP/10/out/kernel-tmp/arch/arm64/boot/dts/qcom/sdm845-tama-akari_generic-overlay.dtbo
```
```
ls -lha /home/developer/android/rom/sonyAOSP/10/out/kernel-tmp/arch/arm64/boot/dts/qcom/sdm845-v2-tama-akari_generic-overlay.dtbo
-rw-r--r-- 1 martin martin 291K Nov 22 18:05 /home/developer/android/rom/sonyAOSP/10/out/kernel-tmp/arch/arm64/boot/dts/qcom/sdm845-v2-tama-akari_generic-overlay.dtbo
```

Thaodan commented:
Please read the log files there's a build error probably.

MartinX3 commented:
I don't see a build error in the build_log_akari
[build_log_akari.txt](https://github.com/sonyxperiadev/kernel-sony-msm-4.14-common/files/3882248/build_log_akari.txt)
The kernel also got created and is 14.7 mb big.

Is there another build log in kernel-tmp I don't see @Thaodan?

And you can build the tama platform correctly with your PR?

And I was able to build the tama kernels before, only with the Marijn commits in his repository.

MarijnS95 commented:
Hi @thaodan

I'm a fan of this PR but unfortunately it looks like it has stalled. I've taken the liberty to revive it, including the following changes:

- Apply review comments (deduplicating the clang script)
- Ensuring `ANDROID_ROOT` is set in all clang.sh and gcc.sh
- Rebase on top of sonyxperiadev:aosp/LA.UM.7.1.r1 to get rid of the conflicts
- Compile all device kernels to their own directory
- Only delete output directories when `CLEAN_AFTER_BUILD=true` (parent dirs are not deleted, not sure if we want to do that at the end. That would delete all device dirs even when the script is edited to build a single device)
- Show `make` output directly on screen, except when `LOG_TO_FILE=true`
- Allow `BUILD_ARGS` to contain a path with spaces, generally used for `CROSS_COMPILE="ccache gcc"`
- Shellcheck warnings are a thing of the past now

You can find it here:
https://github.com/MarijnS95/kernel-sony-msm-4.14-common/commits/massive-cleanup

Do you want to take over my branch (except the `KEEPLOCAL` commits, those are just for example), eventually amend some and update this PR, or shall I PR it?

After this we can look into building single devices/platforms like in #4. Though I would prefer `./build-kernels-gcc.sh griffin` instead of `DEVICE=griffin PLATFORM(s)=kumano ./build-kernels-gcc.sh` but that's a discussion for the other PR :)

Thaodan commented:
Hey

I'll take most of the commits but I don't think its nessary to a dir per compiler 
and device unless you want to support parallel builds.
The output directory shouldn't be deleted to increase the speeds of rebuilds 
when testing changes.

Abou the situatiobn when it comes to logging: it depends on what should be 
the default.
Currently the default was that logs didn't appear on screen.
I'd rather use command line arguments to direct the output of those outputs 
or to set options in general when it comes to that matter.

jerpelea commented:
please do cleanup by default
if you build all kernels it is a huge waste of space

MarijnS95 commented:
> I'll take most of the commits but I don't think its nessary to a dir per compiler
and device unless you want to support parallel builds.

I generally do parallel builds with both compilers so that's a necessity for me when preserving the output (incremental build speed...)

> The output directory shouldn't be deleted to increase the speeds of rebuilds
when testing changes.

We'll keep that option for Alin though, who prefers (understandably) clean builds. He has an insane build machine after all, and can run this safely in the background without too much time constaints.


> Abou the situatiobn when it comes to logging: it depends on what should be the default. Currently the default was that logs didn't appear on screen.

I personally prefer to have it on by default. If you don't notice that the command terminated or are testing locally, having the error pop up on the screen directly is invaluable.
@jerpelea I think you use this to check build outputs for separate devices after the fact, when no file popped up?

> I'd rather use command line arguments to direct the output of those outputs or to set options in general when it comes to that matter.

Like a filename? Sure, we can do that. Same for the delete-after-build I guess.

MarijnS95 commented:
@Thaodan We'll have to remove the `"` around `"$(find "$KERNEL_TMP"/arch/arm64/boot/dts -name "*.dtbo")"` since there are devices like Akatsuki with multiple DTBO files. This causes them to be one large filename with spaces, which is obviously not correct.

We could add something like `-printf "\"%p\" "` (note the space after the `\"`) or do you have a better idea?

Thaodan commented:
Find -print0 could work better.

MarijnS95 commented:
> Find -print0 could work better.

Unfortunately that does not address the issue:
```
$ ls                                
'foo bar.dtbo'   'second file.dtbo'
$ ls $(find -iname '*.dtbo' -print0)  
ls: cannot access './foo': No such file or directory
ls: cannot access 'bar.dtbo': No such file or directory
ls: cannot access './second': No such file or directory
ls: cannot access 'file.dtbo': No such file or directory
ls: cannot access '': No such file or directory
$ ls "$(find -iname '*.dtbo' -print0)"
'./foo bar.dtbo'
```

jerpelea commented:
please rebase

Thaodan committed:
Add proper shebang for files that need to be directly executed

Signed-off-by: Björn Bidar <theodorstormgrade@gmail.com>

Thaodan committed:
Reuse $ANDROID_BUILD_TOP if possible.

Thanks @MarijnS95 for the idea.
Needs realpath from coreutils

Signed-off-by: Björn Bidar <theodorstormgrade@gmail.com>

Thaodan committed:
Separate build directories per build script

Like @MarijnS95's idea but simpler.

Signed-off-by: Björn Bidar <theodorstormgrade@gmail.com>

Thaodan committed:
Add option to keep $KERNEL_TMP

Signed-off-by: Björn Bidar <theodorstormgrade@gmail.com>

Thaodan committed:
Add argument to show usage and show usage on wrong argument

Signed-off-by: Björn Bidar <theodorstormgrade@gmail.com>

Thaodan committed:
No need to export anything, no need for unset.

Signed-off-by: Björn Bidar <theodorstormgrade@gmail.com>

Thaodan committed:
Add argument to build the kernel only for a specific device

Also keep kernel build dir when doing this.

Signed-off-by: Björn Bidar <theodorstormgrade@gmail.com>

Thaodan committed:
Remove $BUILD var and use $BUILD_ARGS to pass additional arguments

Signed-off-by: Björn Bidar <theodorstormgrade@gmail.com>

Thaodan committed:
Drop "$CP_BLOB" as its redundant

Signed-off-by: Björn Bidar <theodorstormgrade@gmail.com>

Thaodan commented:
Hey,

I rebased the branch and added changes made by @MarijnS95  or took inspiration from him.
Also I added the feature requested by @oshmoun to build for a single device by using -d <device> as an argument.
The build dir is only kept when using -k or building for a single device.

Thaodan committed:
Don't quote the output of find to support multiple arguments.

When creating the dtbo image we want word splitting to properly pass
all dtbo files to mkdtimg.
See here for an example:
https://github.com/koalaman/shellcheck/wiki/SC2046#exceptions

Signed-off-by: Björn Bidar <theodorstormgrade@gmail.com>

Thaodan commented:
About the issue when having dtbo files with spaces, please see my last commit. The issue was the was no word splitting for the output of find.

pablomh commented:
Please, consider dropping the "$VARIABLE"/foo and adopt the ${VARIABLE}/foo one (it makes my eyes bleed :( ).

Thaodan commented:
{VAR}/foo doesn't do quoting 
the first one does.

MarijnS95 commented:
Pity you didn't manage to take anything from my branch. Means the time I spent amending fixes to (my own) commits, checking shellcheck and writing proper commit messages/descriptions splitting the features... has all gone to waste.

> About the issue when having dtbo files with spaces, please see my last commit. The issue was the was no word splitting for the output of find.

We want word splitting between the files that `find` emits; in my opinion shellcheck warns for splitting between file paths, since `find` doesn't emit anything properly quoted/escaped on its own.

Code review follows inline... after cooking dinner.

MarijnS95 commented on build_shared.sh:
```diff
@@ -0,0 +1,79 @@
+                echo "================================================="
+                echo "Platform -> ${platform} :: Device -> $device"
+                make O=$KERNEL_TMP ARCH=arm64 CROSS_COMPILE=$CROSS_COMPILE -j$(nproc) ${BUILD_ARGS} aosp_$platform"_"$device\_defconfig \
+                     > aosp_$platform"_"$device\_defconfig 2>&1
```
I don't think piping to `aosp_$platform"_"$device\_defconfig` in the current (`"$KERNEL_TOP"/kernel`) directory is intended here?

Thaodan replied:
This was a leftover from my edits before I'll clean this up.

MarijnS95 commented on build-kernels-clang.sh:
```diff
@@ -1,111 +1,14 @@
-# Cross Compiler
-export GCC_CC=../../prebuilts/gcc/linux-x86/aarch64/aarch64-linux-android-4.9/bin/aarch64-linux-android-
-export CLANG_CC=../../prebuilts/clang/host/linux-x86/clang-r353983c/bin/clang
+. "$OLDPWD"/build_shared_vars.sh
```
Is there a particular reason for quoting just the variable (not only here), instead of writing `"$OLDPWD/build_shared_vars.sh"`? I find that variant much easier on the eyes, as it shows the entire thing is part of the same string.

Also, IMO the script should be made location-invariant. @ix5 already mentioned (and I agree) that $OLDPWD is a bit too much magic in that it depends on your cd commands. This is going to bite someone in the future when changing the script).
Instead, `"$(dirname $0)"` is the way to go IMO. That way we can consistently source the other shell files and know exactly where to put the output. No matter if someone calls `./kernel/sony/msm-4.14/common-kernel/build-xxx.sh`, `./build-xxx.sh` or whatever.

Thaodan replied:
Fixed in https://github.com/sonyxperiadev/kernel-sony-msm-4.14-common/pull/6/commits/a021c7fd9aa29ff6a59a7bb97294fc7822a9c5bf

Thaodan commented:
>Pity you didn't manage to take anything from my branch. Means the time I 
>spent amending fixes to (my own) commits, checking shellcheck and writing 
>proper commit messages/descriptions splitting the features... has all gone to 
>waste.
I did but didn't take the exact commits as I want to make these changes in a 
different order or way. You had for example had two changes in some commits, one that was explained in the commit and one that wasn't.
I added credit to every change I have taken from your branch to make this 
clear.

Shellcheck throws a (yellow) warning in this case as you will the in my 
screenshot in the next message.

Thaodan commented:
<img width="956" alt="Screenshot_20191216_173827" src="https://user-images.githubusercontent.com/1747359/70925118-ff2bd780-202a-11ea-8026-11303231efc6.png">

MarijnS95 commented on build-kernels-clang.sh:
```diff
@@ -1,111 +1,14 @@
+#!/bin/sh
+
 cd ../../../..
```
Now that you have `ANDROID_ROOT`/`ANDROID_BUILD_TOP`, is there a particular reason to not use that here, or get rid of the current directory dependency completely?

Thaodan replied:
Fixed in https://github.com/sonyxperiadev/kernel-sony-msm-4.14-common/pull/6/commits/a021c7fd9aa29ff6a59a7bb97294fc7822a9c5bf

MarijnS95 commented on build-kernels-gcc.sh:
```diff
@@ -1,106 +1,8 @@
-    echo "Building new kernel image ..."
-    echo "Loggin to $KERNEL_TMP/build_log_${device}"
-    $BUILD >$KERNEL_TMP/build_log_${device} 2>&1;
+cd ../../../..
```
Same here, can we get rid of the dependency on where pwd is at?

Thaodan replied:
Fixed in https://github.com/sonyxperiadev/kernel-sony-msm-4.14-common/pull/6/commits/a021c7fd9aa29ff6a59a7bb97294fc7822a9c5bf

MarijnS95 commented on build_shared_vars.sh:
```diff
@@ -0,0 +1,48 @@
+
+KERNEL_TOP=$ANDROID_ROOT/kernel/sony/msm-4.14
+# $KERNEL_TMP sub dir per script
+KERNEL_TMP=$ANDROID_ROOT/out/${0##*/}/kernel-tmp
```
Didn't you intend something like `${${0##*-}%%.sh}`?

Thaodan replied:
I can do it this way to. Looks a bit cleaner.

MarijnS95 commented on build_shared.sh:
```diff
@@ -0,0 +1,79 @@
+
+                echo "================================================="
+                echo "Platform -> ${platform} :: Device -> $device"
+                make O=$KERNEL_TMP ARCH=arm64 CROSS_COMPILE=$CROSS_COMPILE -j$(nproc) ${BUILD_ARGS} aosp_$platform"_"$device\_defconfig \
```
`aosp_$platform"_"$device\_defconfig` should at least use the same escape sequence for `_`, or take advantage of curly braces.

Thaodan replied:
I've only changed belonged to my intended changes but I can fix this part to and it.

Thaodan replied:
Fixed in https://github.com/sonyxperiadev/kernel-sony-msm-4.14-common/pull/6/commits/8fbd2ae85db3113b06ddb2fada972c996cea37f1

MarijnS95 commented on build-kernels-clang.sh:
```diff
@@ -1,111 +1,14 @@
-    fi
-done
-done
+BUILD_ARGS="CLANG_TRIPLE=aarch64-linux-gnu CROSS_COMPILE=$GCC_CC -j$(nproc) CC=$CLANG_CC"
```
`-j$(nproc)` is duplicate, and `CROSS_COMPILE=$GCC_CC` is duplicate too after following `CROSS_COMPILE="$GCC_CC"` down to the make command with `CROSS_COMPILE=$CROSS_COMPILE`.

MarijnS95 replied:
Arguments like `$GCC_CC` and `$_CLANG_CC` should ultimately be escaped, in the event that `ANDROID_ROOT` contains spaces (`$HOME/my android build folder/prebuilts/...`). This does however require `eval` to be parsed properly, see https://github.com/MarijnS95/kernel-sony-msm-4.14-common/commit/a8ffdab0704aa58456f2f334edbe3b9bbf45094f.

MarijnS95 commented:
> I did but didn't take the exact commits as I want to make these changes in a different order or way.

IMO it's easier to leave commits out and amend small changes to others; that way attribution is preserved just like I did with your commits. Fair though, maybe I'm a bit too pedantic when it comes to maintaining authorship for such small changes?
More importantly I'm missing the cleanups that I amended to my own _messy and **not upstreamable**_ commits, and cleanups to the general script and/or amended to yours.

> You had for example had two changes in some commits, one that was explained in the commit and one that wasn't. I added credit to every change I have taken from your branch to make this clear. 

Interesting, I took (extra) care to keep it clean and have every change separated. Guess it's `Build all device kernels to their own directory` while also adding the option to clean after build (which was/is inherently intertwined with device separation)?

MarijnS95 reviewed (changes requested):
Is there a particular reason to mix `$ARGNAME` and `${ARGNAME}` (besides those persisted from the original script)? The latter one doesn't seem to be exclusively used where necessary. For example (this is a remnant from the original version - no blame on you) `aosp_$platform"_"$device\_defconfig` should IMO be written as `"aosp_${platform}_${device}_defconfig"` to take advantage of curly braces notation - making it much more readable at the same time.

On the other hand of the spectrum we have `mkdir -p "${KERNEL_TMP}"`. I don't care which method is used, as long as:
1. There's a valid reason to use one over the other
2. It is consistent across the entire script.

Furthermore a bunch of cleanups have been missed in the picking process.

jerpelea commented:
please update the PR

Thaodan committed:
Use variable substitution/dirname with 0 instead of cd&& $OLDPWD

It is cleaner and makes the first cd redundant.
As there's no cd into the $ANDROID_ROOT first we have to find it by
hand first. Because this project is to be run in an android repo tree
we can assume that the directory that contains /.repo/ is our $ANDROID_ROOT.

Signed-off-by: Björn Bidar <theodorstormgrade@gmail.com>

Thaodan committed:
Make $CLANG_CC more generic and detect if $CC is set in build_shared

This allows safer quoting for $(CLANG_)CC without using eval and
allows the use override of the compiler without a variable that can
be only used by clang.

Signed-off-by: Björn Bidar <theodorstormgrade@gmail.com>

Thaodan committed:
Use consistent variable substitution

If a variable is in argument containing other words use ${foo}bar to
mask it.

Signed-off-by: Björn Bidar <theodorstormgrade@gmail.com>

Thaodan reviewed (commented):
Please review again I linked all commits were I fixed the mentioned issues.
Some issue were error in a commit and I edit the particular commit.

MarijnS95 commented on build_shared.sh:
```diff
@@ -0,0 +1,78 @@
+                echo "The build may take up to 10 minutes. Please be patient ..."
+                echo "Building new kernel image ..."
+                echo "Logging to $KERNEL_TMP/build.log"
+                make O="$KERNEL_TMP" ARCH=arm64 CROSS_COMPILE="$CROSS_COMPILE" -j$(nproc) ${BUILD_ARGS} ${CC+"${CC}"} \
```
Nice substitution rule :)

(Perhaps you want to use `${CC:+"..."}` which will _not_ substitute the rhs when CC is unset _or_ null (`CC=`).

This is missing the argument name to pass to make. Should be `${CC:+CC="${CC}"}` :grin: 

And indeed, as per the description in your comment, passing `CC="ccache gcc"` with the space is much easier/obvious to escape than `BUILD_ARGS="CC=\"ccache gcc\""`.

Thaodan replied:
Thanks for pointing that out fixed it as you suggested. Using ${foo:+bar} looks correcter.

MarijnS95 commented on build_shared_vars.sh:
```diff
@@ -0,0 +1,63 @@
+find_repo_root()
```
I wasn't intending to go "this far" with root detection, just a `$(realpath ../../../..)` would have been enough for me if `ANDROID_BUILD_TOP` isn't set, but such a function is definitely a nice addition.
Anyway I don't like the arbitrary depth number, when a path is guaranteed to have a root somewhere. Nor `cd` ing, but that's mostly personal preference.

Let's pull a concise version from stackoverflow:
https://stackoverflow.com/a/35513635/2844473
```bash
rtrav() { test -e $2/$1 && echo $2 || { test $2 != / && rtrav $1 `dirname $2`;}; }
```
Replacing the `dirname` invocation with `${2%/*}` (since dirname doesn't do any canonicalization, just _string_ manipulations), and perhaps `test` with `[ ]` or `[[ ]]` seems fine to me.
```bash
rtrav() { [ -e $2/$1 ] && echo $2 || { [ $2 != / ] && rtrav $1 ${2%/*};}; }
```
All that's left is adding quotes to prevent globbing and perhaps deal with SC2015 if we ever expect `echo` to fail.

Thaodan replied:
> I wasn't intending to go "this far" with root detection, just a `$(realpath ../../../..)` would have been enough for me if `ANDROID_BUILD_TOP` isn't set, but such a function is definitely a nice addition.
> Anyway I don't like the arbitrary depth number, when a path is guaranteed to have a root somewhere. Nor `cd` ing, but that's mostly personal preference.

I added a limit to prefent the function from jumping outside of the repo tree if its broken into another one (I have to for example)

Thaodan replied:
> Replacing the `dirname` invocation with `${2%/*}` (since dirname doesn't do any canonicalization, just _string_ manipulations), and perhaps `test` with `[ ]` or `[[ ]]` seems fine to me.

Do you think there are cases were ${0} is either not something like ./path/to/script or ./script?
As long as the path is absolute its not important.

MarijnS95 replied:
> I added a limit to prefent the function from jumping outside of the repo tree if its broken into another one (I have to for example)

Then it should anyway find the nested folder containing `.repo`, otherwise different problems arise. 10 folders is a lot, yet arbitrary.

> Do you think there are cases were ${0} is either not something like ./path/to/script or ./script?
> As long as the path is absolute its not important.

`"$PWD"` is always set to an absolute path afaik.

Thaodan replied:
But I was talking about $0 thats what you were talking about. I think this simple solution  is enough here.

Thaodan replied:
>Then it should anyway find the nested folder containing .repo, otherwise different problems arise. 10 folders is a lot, yet arbitrary.

It is arbitrary as I think there needs to be a limit that makes sense but also isn't limited when the depth of the script in the tree changes.

MarijnS95 replied:
`${2%/*}` instead of calling `$(dirname $2)`? I think either is fine. Try `$(dirname .)` or `$(dirname ..)` or `$(dirname ../../..)`, you'll see what I mean :)

MarijnS95 commented on build_shared.sh:
```diff
@@ -0,0 +1,78 @@
+                echo "Copying new kernel image ..."
+                cp "$KERNEL_TMP/arch/arm64/boot/Image.gz-dtb" "$KERNEL_TOP/common-kernel/kernel-dtb-$device"
+                if [ $DTBO = "true" ]; then
+                    # shellcheck disable=SC2046
```
Note that as we discussed earlier, we only want _partial_ wordsplitting here. That is, the paths coming from find need to be separated with a space _BUT_ the individual paths should be quoted, may we ever run into an `arch/arm64/boot/dts/evil dtbo with spaces.dtbo`.

What do you think?

Adding `-printf '"%f"\n'` takes care of this, then we can safely disable the (this time really) false positive.

Thaodan replied:
I know that thats why I disabled the shellcheck warning. I tested script with akari and everything was fine. I can try to add printf to the find command and test.
I find dev environment really an edge case that its ok to support but really its time consuming.

MarijnS95 replied:
That would be nice, just to be sure we aren't running into the (very unlikely) example of a dtbo with spaces.

If you are stressed for time you can always hand the PR over to someone else, while jumping in on the discussion. But to be fair, this one is a massive nitpick :)

Thaodan replied:
>If you are stressed for time you can always hand the PR over to someone else, while jumping in on the discussion. But to be fair, this one is a massive nitpick :)

No problem. I just don't like fixing stupid edge cases when they make things ugly like using eval which is something I try to avoid.
It's normal that such rewrites take time.

MarijnS95 reviewed (approved):
Thanks a lot for applying the review comments, it looks nice and tidy this way :)

Looks like a `CC=` went missing, the rest of the comments are just style/simplicity/correctness.

MartinX3 commented:
Please rebase

jerpelea commented:
@Thaodan please rebase

Thaodan commented:
@jerpelea Will do that today.

Thaodan commented:
Done.

Thaodan committed:
Indent, handle CROSS_COMPILE better

Signed-off-by: Björn Bidar <theodorstormgrade@gmail.com>
